### PR TITLE
[LinalgToXeGPU] Lower `linalg.matmul_transpose_b` into `xegpu.dpas`

### LINKED ISSUE #340: [`LinalgToXeGPU`] Support conversion for `linalg.matmul` with `transpose_b`
This can be implemented [using `xegpu.load_nd %x {transpose = {1, 0}}`](https://github.com/intel/mlir-extensions/blob/3ffa8bfdd6f6ab410efaed5d595b189266979078/test/Integration/Dialect/XeGPU/gemm_with_transposed_B_1kx1kx1k_f16_f16_f32.mlir#L28-L47) for B tiles.

We should support both patterns:
1. Explicit transpose in op `linalg.matmul_transpose_b`.
2. Transpose before `matmul` (this is how MLIR from OV [will look like](https://github.com/slyalin/openvino/pull/167/files#diff-6f3354ea6a4130f6ce53de41972edce4d385d12309099f9574ec8fa6e3791cccR33)):
    ```mlir
    %b_tr = linalg.transpose %b ...
    %res = linalg.matmul %a, %b, ...
    ````

This functionality is required for OV integration (#207).

### PR DESCRIPTION
**Pull request by dchigarev**
Closes #340

Support lowering of `linalg.matmul_transpose_b` as well as `linalg.transpose %b + linalg.matmul %a %b` into `xegpu.dpas`.

To make a transposed multiplication we have to load B chunks with `xegpu.load_nd ... <transpose = array<i64: 1, 0>>` attribute and also change the iteration dimension (from rows to cols).

### CODE REVIEW

File: lib/gc/Transforms/GPU/LinalgToXeGPU.cpp [hunk 1]
```diff
@@ -669,6 +669,9 @@ static SmallVector<Value> createDescriptorTiles(PatternRewriter &rewriter,
     Value newRowOffs = rewriter.create<arith::ConstantIndexOp>(loc, i);
     for (int j = 0; j < loadShape[1]; j += descTile[1] * arrayLength) {
       Value newColOffs = rewriter.create<arith::ConstantIndexOp>(loc, j);
+      if (transpose) {
+        std::swap(newRowOffs, newColOffs);
```
Comment (dchigarev) [inline lib/gc/Transforms/GPU/LinalgToXeGPU.cpp:673]: changes iteration dimension for B chunks

File: lib/gc/Transforms/GPU/LinalgToXeGPU.cpp [hunk 2]
```diff
@@ -860,13 +865,82 @@ extractVecSubTiles(PatternRewriter &rewriter, Location loc,
...
+            isa<linalg::GenericOp>(trUser)) {
+          auto matmulOp = dyn_cast<linalg::LinalgOp>(trUser);
+          auto actualMatmulOperand = matmulOp.getDpsInputs()[1];
+          if (actualMatmulOperand != matmulOperand) {
+            return rewriter.notifyMatchFailure(
+                trUser,
+                "Transpose result is used by more than one matmul operation");
+          }
+        } else if (isa<memref::DeallocOp>(trUser) ||
+                   isa<memref::AllocaOp>(trUser)) {
```
Comment (kurapov-peter) [inline lib/gc/Transforms/GPU/LinalgToXeGPU.cpp:908]: What would alloca mean?

**Review by kurapov-peter — COMMENTED** [review]
This seem to be almost generic enough to support transpose A and the batched transposed versions. Could you try making the interfaces amenable to all the matmul variations? Implementation can stay as is.

File: lib/gc/Transforms/GPU/LinalgToXeGPU.cpp (hunk 2 above)
Comment (dchigarev) [reply lib/gc/Transforms/GPU/LinalgToXeGPU.cpp:908]: nothing, it was added here by mistake, will remove in the next commit

**Comment by dchigarev** [comment]
> This seem to be almost generic enough to support transpose A and the batched transposed versions. Could you try making the interfaces amenable to all the matmul variations? Implementation can stay as is.

1. Regarding batched transpose: at the moment [there's no support](https://github.com/intel/graph-compiler/blob/6e13abe56632fde403ce770cadae9294f8ac325a/lib/gc/Transforms/GPU/LinalgToXeGPU.cpp#L1494-L1499) for either `linalg.batch_matmul` or for `linalg.batch_reduce_matmul` in `linalg-to-xegpu` pass. Trying YOLO approach by simply adding those ops to the list of allowed ops causes a segfault, so probably a different handling is needed in those cases. Anyway, the 'transpose_b' logic shouldn't change at all if `createDPASKernel()` would support batched matmuls sometime. You would just need to add `BatchMatmulOp` everywhere and it should work.


2. Regarding 'transpose A': this one also requires special handling. Following the current logic of `createDPASKernel()` we can't use `transpose` attribute of `xegpu.load_nd` to transpose A argument since it only can transpose 32bit blocks (and we [only support f16](https://github.com/intel/graph-compiler/blob/6e13abe56632fde403ce770cadae9294f8ac325a/lib/gc/Transforms/GPU/LinalgToXeGPU.cpp#L117-L122) in our lowering). This means, that instead of using `xegpu.load_nd <transpose>` we would need to use `vector.transpose` on the loaded vector. Since it doesn't correlate with the changes in this PR and seems like a separate feature, I would propose to consider it as a separate issue (#361). Anyway, the general logic of detecting "if there's a transpose before matmul" stays the same. Added an argument to `findAndReplaceTranspose()` function indicating which matmul operand we're processing.
    <details><summary>Example of why 32bit transpose causes problems for f16 data</summary>
    
    ```
    A(f16): // original matrix
    [1,   2,  3,  4]
    [5,   6,  7,  8]
    [9,  10, 11, 12]
    [13, 14, 15, 16]
    
    xegpu.load_nd A(f16) <transpose>: // transposed 32bit blocks
    [1, 2,  9, 10]
    [5, 6, 13, 14]
    [3, 4, 11, 12]
    [7, 8, 15, 16]
    
    vector.transpose A(f16): // expected transpose result
    [1, 5,  9, 13]
    [2, 6, 10, 14]
    [3, 7, 11, 15]
    [4, 8, 12, 16]
    ```
    
    </details>

    <details><summary>Why it doesn't cause any problems for B argument</summary>
     
    In contrast with 'A' argument we load 'B' with 'packed/vnni' attribute which apparently involves [some special handling](https://github.com/intel/mlir-extensions/blob/595e81763e4b450b491b206e5f23854d2fa05193/docs/rfcs/XeGPU.md?plain=1#L142) which is properly aligned with our needs of B transpose.
    
    </details>

    P.S. there's [`transpose_bit_width`](https://github.com/intel/mlir-extensions/blob/595e81763e4b450b491b206e5f23854d2fa05193/docs/rfcs/XeGPU.md?plain=1#L135) attribute that should in theory solve this problem, but setting it to 16 does nothing. Not sure if this an intended behavior or just a bug in `xegpu-to-vc-func` lowering. Submitted a question to IMEX (intel/mlir-extensions#895).

**Comment by kurapov-peter** [comment]
> > This seem to be almost generic enough to support transpose A and the batched transposed versions. Could you try making the interfaces amenable to all the matmul variations? Implementation can stay as is.
> 
> 1. Regarding batched transpose: at the moment [there's no support](https://github.com/intel/graph-compiler/blob/6e13abe56632fde403ce770cadae9294f8ac325a/lib/gc/Transforms/GPU/LinalgToXeGPU.cpp#L1494-L1499) for either `linalg.batch_matmul` or for `linalg.batch_reduce_matmul` in `linalg-to-xegpu` pass. Trying YOLO approach by simply adding those ops to the list of allowed ops causes a segfault, so probably a different handling is needed in those cases. Anyway, the 'transpose_b' logic shouldn't change at all if `createDPASKernel()` would support batched matmuls sometime. You would just need to add `BatchMatmulOp` everywhere and it should work.
> 2. Regarding 'transpose A': this one also requires special handling. Following the current logic of `createDPASKernel()` we can't use `transpose` attribute of `xegpu.load_nd` to transpose A argument since it only can transpose 32bit blocks (and we [only support f16](https://github.com/intel/graph-compiler/blob/6e13abe56632fde403ce770cadae9294f8ac325a/lib/gc/Transforms/GPU/LinalgToXeGPU.cpp#L117-L122) in our lowering). This means, that instead of using `xegpu.load_nd <transpose>` we would need to use `vector.transpose` on the loaded vector. Since it doesn't correlate with the changes in this PR and seems like a separate feature, I would propose to consider it as a separate issue ([[LinalgToXeGPU] Support `linalg.matmul_transpose_a` #361](https://github.com/intel/graph-compiler/issues/361)). Anyway, the general logic of detecting "if there's a transpose before matmul" stays the same. Added an argument to `findAndReplaceTranspose()` function indicating which matmul operand we're processing.
>     Example of why 32bit transpose causes problems for f16 data
>    ```
>    A(f16): // original matrix
>    [1,   2,  3,  4]
>    [5,   6,  7,  8]
>    [9,  10, 11, 12]
>    [13, 14, 15, 16]
>    
>    xegpu.load_nd A(f16) <transpose>: // transposed 32bit blocks
>    [1, 2,  9, 10]
>    [5, 6, 13, 14]
>    [3, 4, 11, 12]
>    [7, 8, 15, 16]
>    
>    vector.transpose A(f16): // expected transpose result
>    [1, 5,  9, 13]
>    [2, 6, 10, 14]
>    [3, 7, 11, 15]
>    [4, 8, 12, 16]
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>     
>     Why it doesn't cause any problems for B argument
>    In contrast with 'A' argument we load 'B' with 'packed/vnni' attribute which apparently involves [some special handling](https://github.com/intel/mlir-extensions/blob/595e81763e4b450b491b206e5f23854d2fa05193/docs/rfcs/XeGPU.md?plain=1#L142) which is properly aligned with our needs of B transpose.
>     
>    P.S. there's [`transpose_bit_width`](https://github.com/intel/mlir-extensions/blob/595e81763e4b450b491b206e5f23854d2fa05193/docs/rfcs/XeGPU.md?plain=1#L135) attribute that should in theory solve this problem, but setting it to 16 does nothing. Not sure if this an intended behavior or just a bug in `xegpu-to-vc-func` lowering. Submitted a question to IMEX ([[xegpu-to-vc-func] Is `transpose_bit_width=16` supported? mlir-extensions#895](https://github.com/intel/mlir-extensions/issues/895)).

Sure, I did not suggest adding support for other flavors right away, only to make the interfaces friendly. Supporting other matmuls is a separate issue.

**Review by kurapov-peter — APPROVED** [review]
There's also the bool transpose param in a couple places that could be generalized but that's a minor thing.

**Comment by dchigarev** [comment]
> There's also the bool transpose param in a couple places that could be generalized but that's a minor thing.

This bool parameter is used in functions that also take a matmul operand, so this parameter can be used with both of the operands (and with any matmul op). Is this generic enough or you had something else on your mind?

```cpp
bool transposeA = true;
bool transposeB = false;

auto tilesA = createTiles(matA, ..., /*transpose=*/transposeA);
auto tilesB = createTiles(matB, ..., /*transpose=*/transposeB);
```

**Comment by kurapov-peter** [comment]
> > There's also the bool transpose param in a couple places that could be generalized but that's a minor thing.
> 
> 
> 
> This bool parameter is used in functions that also take a matmul operand, so this parameter can be used with both of the operands (and with any matmul op). Is this generic enough or you had something else on your mind?
> 
> 
> 
> ```cpp
> 
> bool transposeA = true;
> 
> bool transposeB = false;
> 
> 
> 
> auto tilesA = createTiles(matA, ..., /*transpose=*/transposeA);
> 
> auto tilesB = createTiles(matB, ..., /*transpose=*/transposeB);
> 
> ```

Yup. Sounds good!

**Review by AndreyPavlenko — APPROVED** [review]

**Pull request merged**